File: frontend/src/api/ai.ts
import request from '@/utils/request'
import axios from 'axios'
import { getApiUrl } from '@/utils/request'
import type { Image } from '@/types/image'
import type { Software } from '@/types/software'

interface AIResponse {
  result: string;
}

interface GenerateDescriptionRequest {
  image: {
    name: string;
    version: string;
    architecture: string;
  };
  software_list: Array<{
    name: string;
    version: string;
    architecture: string;
  }>;
}

interface OptimizeDockerfileRequest {
  dockerfile: string;
}

interface CompatibilityCheckRequest {
  base_image: {
    name: string;
    version: string;
    architecture: string;
  };
  software_list: Array<{
    name: string;
    version: string;
    architecture: string;
  }>;
}

export const generateDescription = async (data: GenerateDescriptionRequest): Promise<string> => {
  try {
    const response = await request({
      url: getApiUrl('/ai/generate_description'),
      method: 'post',
      data
    })
    
    // 处理不同格式的响应
    if (typeof response === 'object') {
      // 如果响应是一个对象，尝试获取description或result字段
      if (response.description) {
        return response.description
      } else if (response.result) {
        return response.result
      } else if (typeof response === 'string') {
        return response
      } else {
        // 将整个响应转换为JSON字符串作为备选
        return JSON.stringify(response)
      }
    } else if (typeof response === 'string') {
      // 如果响应直接是字符串
      return response
    }
    
    // 如果无法识别响应格式，返回空字符串
    console.warn('生成描述返回了意外格式:', response)
    return ''
  } catch (error) {
    console.error('生成描述错误:', error)
    // 返回空字符串而不是抛出错误
    throw error
  }
}

export const optimizeDockerfile = async (dockerfile: string): Promise<any> => {
  try {
    const request_data: OptimizeDockerfileRequest = {
      dockerfile: dockerfile
    }
    
    const response = await request({
      url: getApiUrl('/ai/optimize_dockerfile'),
      method: 'post',
      data: request_data
    })
    
    // 如果响应数据为undefined，则直接返回响应本身
    if (response === undefined) {
      return response
    }
    
    return response
  } catch (error) {
    console.error('Optimize Dockerfile error:', error)
    throw error
  }
}

export const checkCompatibility = async (image: Image, softwareList: Software[]): Promise<any> => {
  try {
    const request_data: CompatibilityCheckRequest = {
      base_image: {
        name: image.name,
        version: image.version,
        architecture: image.architecture
      },
      software_list: softwareList.map(s => ({
        name: s.name,
        version: s.version,
        architecture: s.architecture
      }))
    }
    
    const response = await request({
      url: getApiUrl('/ai/check_compatibility'),
      method: 'post',
      data: request_data
    })
    
    // 如果响应数据为undefined，则直接返回响应本身
    if (response === undefined) {
      return response
    }
    
    return response
  } catch (error) {
    console.error('Check compatibility error:', error)
    throw error
  }
} 

File: backend/app/api/ai.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import List, Dict, Any
from sqlalchemy.orm import Session
from app.api import deps
from app.models.user import User
from app.models.image import Image
from app.models.software import Software
from app.services.ai_service import generate_description, optimize_dockerfile, check_software_compatibility
import logging
import requests
from app.core.config import settings

logger = logging.getLogger(__name__)
router = APIRouter()

# 请求模型定义
class GenerateDescriptionRequest(BaseModel):
    image: Dict[str, Any]
    software_list: List[Dict[str, Any]]

class OptimizeDockerfileRequest(BaseModel):
    dockerfile: str

class CompatibilityCheckRequest(BaseModel):
    base_image_id: int
    software_ids: List[int]

class GenerateResponse(BaseModel):
    result: str

# API 端点
@router.post("/generate_description", response_model=GenerateResponse)
async def generate_description_api(request: GenerateDescriptionRequest):
    """
    生成环境描述 API 端点
    """
    try:
        result = await generate_description(request.image, request.software_list)
        return {"result": result}
    except Exception as e:
        logger.error(f"Error in generate_description_api: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Error generating description: {str(e)}"
        )

@router.post("/optimize_dockerfile", response_model=GenerateResponse)
async def optimize_dockerfile_api(request: OptimizeDockerfileRequest):
    """
    优化 Dockerfile API 端点
    """
    try:
        result = await optimize_dockerfile(request.dockerfile)
        return {"result": result}
    except Exception as e:
        logger.error(f"Error in optimize_dockerfile_api: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Error optimizing Dockerfile: {str(e)}"
        )

@router.post("/check_compatibility", response_model=GenerateResponse)
async def check_compatibility_api(
    *,
    db: Session = Depends(deps.get_db),
    request: CompatibilityCheckRequest
):
    """检查兼容性"""
    try:
        # 获取基础镜像信息
        base_image = db.query(Image).filter(Image.id == request.base_image_id).first()
        if not base_image:
            raise HTTPException(status_code=404, detail="Base image not found")
            
        # 获取软件列表信息
        software_list = db.query(Software).filter(Software.id.in_(request.software_ids)).all()
        if not software_list:
            raise HTTPException(status_code=404, detail="Software not found")
        
        # 转换为字典格式
        base_image_dict = {
            "name": base_image.name,
            "version": base_image.version,
            "architecture": base_image.architecture
        }
        
        software_list_dict = [{
            "name": sw.name,
            "version": sw.version,
            "architecture": sw.architecture
        } for sw in software_list]
        
        # 调用兼容性检查服务
        result = await check_software_compatibility(base_image_dict, software_list_dict)
        return {"result": result}
        
    except Exception as e:
        logger.error(f"Error in check_compatibility_api: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Error checking compatibility: {str(e)}"
        ) 